### ISSUE TITLE
Weirdness with study tags (extra tags, sometimes hidden)

### ISSUE THREAD

**Issue by jimallman** [OpenTreeOfLife/opentree#562]
Over a number of saves (messing with study tags), I've accumulated [lots of redundant tag values in study ot_54](https://github.com/OpenTreeOfLife/phylesystem-1/blob/d92be19a93efca4f79ee9a2d4202668812fc8edc/study/ot_54/ot_54/ot_54.json#L139-L148)

These appear differently in different views, sometimes for known reasons:

 - In the curation app's main study list, only one tag appears:
![screen shot 2015-01-23 at 9 38 43 pm](https://cloud.githubusercontent.com/assets/446375/5885839/4d0ea5d2-a348-11e4-9534-943e0240c034.png)

 - In the read-only study view, we see them all:
![screen shot 2015-01-23 at 9 38 55 pm](https://cloud.githubusercontent.com/assets/446375/5885841/556a303e-a348-11e4-8bc6-7ce24df1685a.png)

 - In the study editor, redundant tag values are hidden (but not consolidated when we save):
![screen shot 2015-01-23 at 9 39 06 pm](https://cloud.githubusercontent.com/assets/446375/5885843/5a3a68d6-a348-11e4-85af-b8502bdbe625.png)

It's most important that we consolidate duplicate tags when saving. But first let's clean up the read-only view to hide dupes, and the main study list to show all unique tags.

**Comment by jimallman** [comment]
There's an [oti bug](https://github.com/OpenTreeOfLife/oti/issues/35) that causes only one tag per study to appear in the main study list. (It appears only one tag value is being indexed per study.)

**Comment by jimallman** [comment]
Regarding the accumulation of "delete me" tags, the culprit seems to be a combination of
 - the bootstrap-tagsinput widget and
 - multi-word tag text.

Specifically, the tags widget reads OPTION values from a chosen SELECT widget, and scans its own TagsInput object for matching values. But for simple string values, it messes things up:

When [it first scans their `value` attributes], the values are read "directly" without transformation. So for this OPTION, the value is read as "delete me":
```html
    <option value="delete me">delete me</option>
```
But when it looks for matching OPTION elements (to decide whether a new one needs to be created), [it checks for an URL-encoded version of this value](https://github.com/TimSchlechter/bootstrap-tagsinput/blob/0.3.9/src/bootstrap-tagsinput.js#L112). This would only match on an OPTION like:
```html
    <option value="delete%20me">delete me</option>
```
If we try to pre-escape the `value` attributes, we'll start with OPTIONSs like the one above. But then these are scanned as initial tag values, and the test ends up looking for this:
```html
    <option value="delete%2520me">delete me</option>
```
So we can't win. Perhaps this was fixed in a more recent release? If not, I'll need to fork this or switch from simple string tag values to a simple object like
```json
    {"text": "delete me", "value": "delete%20me" }
```
...just to use other code paths and skip the SELECT and OPTION elements as stores of value.

**Comment by jar398** [comment]
Am I to understand that we didn't write the tags widget, and that it
contains a bug ("we can't win" = it doesn't handle tags containing spaces)?

I'm tempted to say that our tags should be like those in twitter, facebook,
pinboard, etc. - they don't contain spaces. Perhaps this is what the widget
is expecting too. This is also relevant to 'tag clouds' where you wouldn't
be able to distinguish one tag with a space from two tags that happened to
land next to one another.

On the other hand, labels in gmail and github can have spaces, so I don't
know. Forking a widget sounds almost obscene.

**Comment by jimallman** [comment]
I'd like to preserve multi-word tags, since there's no reason for us to restrict them. Besides, the same bug will rear its head once we have a tag like "invalid?" or "support > 0.5". We'll see this behavior whenever `tagValue !== encodeURIComponent( tagValue )`.

> Am I to understand that we didn't write the tags widget, and that it contains a bug ("we can't win" = it doesn't handle tags containing spaces)?

Yes, but the bug only appears when using a particular combination of HTML elements and value types. We use an old release of the tags widget, but I've checked the latest version and the bug survives to this day. I'll report it to that project, but for our purposes the easiest workaround is to use this widget differently. 

By which I mean attaching it to an INPUT rather than a SELECT element, or passing objects (vs. strings) as tag values. This should avoid the bad code -- I just need to make the change and test thoroughly to make sure we never lose tag data.

**Comment by jimallman** [comment]
This is fixed and working now on **devtree**. I kept the SELECT elements for tags, but (as discussed above) using simple objects instead of string values, and populating the widget via Javascript instead of using OPTION tags. This is working well. I've also added code to purge duplicate tags, so we'll gradually clean up any studies with "delete me" and other values that are changed by JS `escape()`.

I've also tested this solution with tags including question marks, angle brackets, and double quotes. All are encoded properly and stored intact in the study nexson.
